osu.Game/Database/LegacyModelExporter.cs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using osu.Framework.Graphics;
+using osu.Framework.Logging;
+using osu.Framework.Platform;
+using osu.Game.Extensions;
+using osu.Game.Overlays;
+using osu.Game.Overlays.Notifications;
+using Realms;
+using SharpCompress.Archives.Zip;
+
+namespace osu.Game.Database
+{
+    /// <summary>
+    /// A class which handles exporting legacy user data of a single type from osu-stable.
+    /// </summary>
+    public abstract class LegacyModelExporter<TModel> : Component
+        where TModel : RealmObject, IHasNamedFiles, IHasGuidPrimaryKey
+    {
+        /// <summary>
+        /// The file extension for exports (including the leading '.').
+        /// </summary>
+        protected abstract string FileExtension { get; }
+
+        protected Storage UserFileStorage;
+        protected Storage ExportStorage;
+
+        protected RealmAccess RealmAccess;
+
+        private readonly ProgressNotification notification;
+
+        private string filename = "";
+
+        private bool canCancel = true;
+
+        protected LegacyModelExporter(Storage storage, RealmAccess realm, INotificationOverlay? notifications = null)
+        {
+            ExportStorage = storage.GetStorageForDirectory(@"exports");
+            UserFileStorage = storage.GetStorageForDirectory(@"files");
+            RealmAccess = realm;
+
+            notification = new ProgressNotification
+            {
+                State = ProgressNotificationState.Active,
+                Text = "Exporting...",
+                CompletionText = "Export completed"
+            };
+            notification.CancelRequested += () => canCancel;
```

Comment:
How does this even work? It's checking whether it.. can cancel but not actually performing any cancellation?

Reply:
I use `notification.CancellationToken.IsCancellationRequested` to confirm whether task is request cancel.

when export zip stream, the task cannot be cancelled, so I use `canCancel` to avoid user cancel the task